core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+            `{
+                type: ... ,
+                payload: ...
+            }`
+        """
+        pass
+
+    def update_state(self, new_state_wrapper: list) -> None:
+        """
+        Given the freshly-downloaded state, call a handler on each transaction
+        that was not already present in our own state
+        """
+        new_state = dict(new_state_wrapper).get('messages')
+        len_state = len(self.state)
+        for i in new_state[len_state:]:
+            self.state.append(i)
+
+    def setter(self, key: str, value: object, flag: bool = False) -> str:
+        """
+        Provided a key and a JSON/np.array object, upload the object to
+        IPFS and then store the hash as the value on the blockchain. The key
+        should be a backward reference to a prior tx
+        """
+        on_chain_value = self._upload(value) if value else None
+        key = on_chain_value if flag else key
+        tx = {'key': key, 'content': on_chain_value}
+        try:
+            tx_receipt = requests.post(
+                "http://localhost:{0}/txs".format(self.port), json=tx)
+            tx_receipt.raise_for_status()
+        except Exception as e:
+            logging.info("HTTP POST error, got: {0}".format(e))
+        return tx_receipt.text
+
+    def getter(self, key: str) -> list:
+        """
+        Next, provided a key, get the IPFS hash
+        from the blockchain and download the object from IPFS
+        """
+        logging.info("Getting from blockchain...")
+        self.update_state(self.get_global_state())
+        retval = self._download(key)
+        return retval
+
+    def _upload(self, obj: object) -> str:
+        """
+        Provided any Python object, store it on IPFS and then upload
+        the hash that will be uploaded to the blockchain as a value
+        """
+        ipfs_hash = self._content_to_ipfs(obj)
+        return str(ipfs_hash)
+
+    def _download(self, key: str) -> object:
+        """
+        Provided an on-chain key, retrieve the value from local state and
+        retrieve the Python object from IPFS
+        TODO: implement a better way to parse through state list
+        """
+        relevant_txs = [self._ipfs_to_content(tx.get('content'))
+                            for tx in self.state if (tx.get('key') == key)]
+        return relevant_txs
+
+    def _ipfs_to_content(self, ipfs_hash: str) -> object:
+        """
+        Helper function to retrieve a Python object from an IPFS hash
+        """
+        return self.client.get_json(ipfs_hash)
+
+    def _content_to_ipfs(self, content: dict) -> str:
+        """
+        Helper function to deploy a Python object onto IPFS, 
+        returns an IPFS hash
+        """
+        ipfs_hash = self.client.add_json(content)
+        return ipfs_hash
+
+    ##########################################################################
+    ###                         DEVELOPER SECTION                          ###
```

Comment:
Please remove the developer section from the provider codebase.

Reply:
need to refactor but we want feedback on this in this PR rather than proposing this as separate PR, getting changes, having to update this, etc

Reply:
Don't assume you will use the same rest of the code on the other codebase. The modules on the other codebase can have completely different requirements from this codebase, and it's out of the scope of this codebase to review this section, so please remove.

Reply:
@georgymh where do we move it to?

Reply:
Remove it from this PR if it's not being used in the Service, and add it in a PR on the Notebook repo if applicable.

Reply:
It's being used in the Service for internal testing. It will be removed as soon as the Notebook can test the Service.